### PR TITLE
Params API: {utf8,octet}_ptr need to know the data size

### PR DESCRIPTION
**Pull request by levitte**
When the purpose is to pass parameters to a setter function, that
setter function needs to know the size of the data passed.  This
remains true for the pointer data types as well.

### CODE REVIEW

**Comment by paulidale** [comment]
In both of these cases, the size will always be the size of the pointer won't it?
If so, there doesn't seems to be a need have it as a parameter -- the various macros and functions could fill it in unbeknownst to the user.

Or have I missed something?

**Comment by levitte** [comment]
From [the `OSSL_PARAM` manual]():

> `data_size` must be set to the size of the data, not the size of the pointer to the data. If this is used in a parameter request, `data_size` is not relevant. However, the responder will set `*return_size` to the size of the data.

The implication is that for *setting* params (i.e. calling a *setter* rather than a *responder*), `data_size` *is* relevant.  You only have to look at `get_ptr_internal` to see it.

Granted, for `utf8_ptr`, it's not really that crucial, as `strlen` can provide the same answer, but it's *absolutely* crucial for `octet_ptr`.  How will the receiving end know the data size otherwise?

**Comment by levitte** [comment]
> If so, there doesn't seems to be a need have it as a parameter -- the various macros and functions could fill it in unbeknownst to the user.

Except that there's no way to correct the 0 that the `OSSL_PARAM_construct_octet_ptr` place in `data_size` except for manipulating it directly, i.e.:

``` C
    params[i] = OSSL_PARAM_construct_octet_ptr("foo", &buf, NULL);
    params[i].data_size = buf_size;
```

File: test/params_test.c [hunk 1]
```diff
@@ -401,7 +401,7 @@ static OSSL_PARAM *construct_api_params(void)
     params[n++] = OSSL_PARAM_construct_utf8_string("p5", app_p5,
                                                    sizeof(app_p5), &app_p5_l);
     params[n++] = OSSL_PARAM_construct_utf8_ptr("p6", (char **)&app_p6,
-                                                &app_p6_l);
+                                                sizeof(app_p6), &app_p6_l);
```
Comment (paulidale) [inline test/params_test.c]: The three places in params_test the app_p6 is defined as a param (`static_raw_params`, `static_api_params` and here), the size is set to sizeof(pointer) which is incorrect if it is meant to contain the length of what app_p6 points at.

Comment (levitte) [reply test/params_test.c]: You're quite right...  and also, there setting test ignores p6 completely...  thanks for pointing it out.

**Comment by levitte** [comment]
Fixed the tests and rebased

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

bbcaef632440067d173e2c4bfc40dd96ef2c0112 test/params_test.c : Adjust tests to check utf8_ptr sizes
f55ed701a458e3b3840a5d8c8dd3019d7d71a26f Params API: {utf8,octet}_ptr need to know the data size

**Pull request closed without merging**